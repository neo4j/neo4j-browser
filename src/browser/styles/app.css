/* Base imports */
@import './base/reset.css';
@import './base/variables.css';
@import './base/fonts.css';

/* Component styles */
@import './components/buttons.css';
@import './components/forms.css';
@import './components/icons.css';

/* Theme styles - loaded dynamically based on theme */
@import './themes/light.css' (prefers-color-scheme: light);
@import './themes/dark.css' (prefers-color-scheme: dark);

/* Essential utility classes that extend Tailwind */
.neo4j-mono {
  font-family: var(--font-mono);
}

.neo4j-sans {
  font-family: var(--font-sans);
}

/* Add any other app-wide styles here */ 