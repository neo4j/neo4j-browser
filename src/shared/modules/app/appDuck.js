/*
 * Copyright (c) 2002-2020 "Neo4j,"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

// Action type constants
export const NAME = 'app'
export const APP_START = `${NAME}/APP_START`
export const USER_CLEAR = `${NAME}/USER_CLEAR`

export const URL_ARGUMENTS_CHANGE = `${NAME}/URL_ARGUMENTS_CHANGE`

// State constants
export const DESKTOP = 'DESKTOP'
export const WEB = 'WEB'
export const CLOUD = 'CLOUD'

const SECURE_SCHEMES = ['neo4j+s', 'bolt+s']
const INSECURE_SCHEMES = ['neo4j', 'bolt']

// Selectors
export const getHostedUrl = state => (state[NAME] || {}).hostedUrl || null
export const getEnv = state => (state[NAME] || {}).env || WEB
export const hasDiscoveryEndpoint = state =>
  [WEB, CLOUD].includes(getEnv(state))
export const inWebEnv = state => getEnv(state) === WEB
export const inWebBrowser = state => [WEB, CLOUD].includes(getEnv(state))
export const getAllowedBoltSchemes = (state, encryptionFlag) => {
  const isHosted = inWebBrowser(state)
  const hostedUrl = getHostedUrl(state)
  return !isHosted
    ? encryptionFlag
      ? SECURE_SCHEMES
      : [...SECURE_SCHEMES, ...INSECURE_SCHEMES]
    : (hostedUrl || '').startsWith('https')
    ? SECURE_SCHEMES
    : INSECURE_SCHEMES
}

// Reducer
export default function reducer(state = { hostedUrl: null }, action) {
  switch (action.type) {
    case APP_START:
      return { ...state, hostedUrl: action.url, env: action.env }
    default:
      return state
  }
}
