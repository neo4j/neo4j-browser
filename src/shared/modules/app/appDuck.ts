/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { GlobalState } from 'shared/globalState'

// Action type constants
export const NAME = 'app'
export const APP_START = 'app/APP_START'
export const UPDATE_BUILD_INFO = 'app/UPDATE_BUILD_INFO'
export const USER_CLEAR = 'app/USER_CLEAR'
export type AppStartAction = { type: typeof APP_START }
export type UserClearAction = { type: typeof USER_CLEAR }

export const URL_ARGUMENTS_CHANGE = 'app/URL_ARGUMENTS_CHANGE'

// State constants
export const DESKTOP = 'DESKTOP'
export const WEB = 'WEB'
export const CLOUD = 'CLOUD'
export type Environment = typeof DESKTOP | typeof WEB | typeof CLOUD

export const SECURE_SCHEMES = ['neo4j+s', 'bolt+s']
export const INSECURE_SCHEMES = ['neo4j', 'bolt']
export const CLOUD_SCHEMES = ['neo4j+s']

// Selectors
export const getHostedUrl = (state: GlobalState): string | null =>
  (state[NAME] || {}).hostedUrl || null
export const getEnv = (state: GlobalState): Environment =>
  (state[NAME] || {}).env || WEB
export const hasDiscoveryEndpoint = (state: GlobalState): boolean =>
  [WEB, CLOUD].includes(getEnv(state))
export const inWebEnv = (state: GlobalState): boolean => getEnv(state) === WEB
export const inCloudEnv = (state: GlobalState): boolean =>
  getEnv(state) === CLOUD
export const inWebBrowser = (state: GlobalState): boolean =>
  [WEB, CLOUD].includes(getEnv(state))
export const inDesktop = (state: GlobalState): boolean =>
  getEnv(state) === DESKTOP

export const getAllowedBoltSchemes = (
  state: GlobalState,
  encryptionFlag?: any
) => {
  const isHosted = inWebBrowser(state)
  const hostedUrl = getHostedUrl(state)
  return !isHosted
    ? encryptionFlag
      ? SECURE_SCHEMES
      : [...SECURE_SCHEMES, ...INSECURE_SCHEMES]
    : (hostedUrl || '').startsWith('https')
    ? SECURE_SCHEMES
    : INSECURE_SCHEMES
}
// currently only Desktop specific
export const isRelateAvailable = (state: GlobalState): boolean =>
  Boolean(
    state[NAME].relateUrl &&
      state[NAME].relateApiToken &&
      state[NAME].relateProjectId
  )
export const getProjectId = (state: GlobalState): string | undefined =>
  state[NAME].relateProjectId

export type AppState = {
  hostedUrl?: string | null
  env?: Environment
  relateUrl?: string
  relateApiToken?: string
  relateProjectId?: string
  neo4jDesktopGraphAppId?: string
}

// Reducer
export default function reducer(
  state: AppState = { hostedUrl: null },
  action: any
): AppState {
  switch (action.type) {
    case APP_START:
      return {
        ...state,
        hostedUrl: action.url,
        env: action.env,
        relateUrl: action.relateUrl,
        relateApiToken: action.relateApiToken,
        relateProjectId: action.relateProjectId,
        neo4jDesktopGraphAppId: action.neo4jDesktopGraphAppId
      }
    default:
      return state
  }
}
