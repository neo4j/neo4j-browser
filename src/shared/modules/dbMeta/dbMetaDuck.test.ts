/*
 * Copyright (c) "Neo4j"
 * Neo4j Sweden AB [http://neo4j.com]
 *
 * This file is part of Neo4j.
 *
 * Neo4j is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
import reducer, {
  CLEAR_META,
  UPDATE_META,
  UPDATE_SERVER,
  UPDATE_SETTINGS
} from './dbMetaDuck'
import { APP_START } from 'shared/modules/app/appDuck'

describe('hydrating state', () => {
  test('should merge inital state and state on load', () => {
    // Given
    const action = { type: APP_START }

    // When
    const hydratedState = reducer({ foo: 'bar' } as any, action)

    // Then
    expect(hydratedState).toMatchSnapshot()
  })

  test('can update server settings', () => {
    // Given
    const initState: any = {
      shouldKeep: true
    }
    const action = {
      type: UPDATE_SETTINGS,
      settings: {
        'browser.test': 1
      }
    }

    // When
    const nextState = reducer(initState, action)

    // Then
    expect(nextState).toMatchSnapshot()
  })

  test('can update server info', () => {
    // Given
    const initState: any = {
      shouldKeep: true
    }
    const action = {
      type: UPDATE_SERVER,
      version: '3.2.0-RC2',
      edition: 'enterprise',
      storeId: 'xxxx'
    }

    // When
    const nextState = reducer(initState, action)

    // Then
    expect(nextState).toMatchSnapshot()
  })

  test('can CLEAR to reset state', () => {
    // Given
    const initState: any = {
      shouldKeep: false,
      server: {
        edition: 'enterprise',
        storeId: 'xxxx',
        version: '3.2.0-RC2'
      }
    }
    const action = {
      type: CLEAR_META
    }

    // When
    const nextState = reducer(initState, action)

    // Then
    expect(nextState).toMatchSnapshot()
  })

  test('can update meta values with UPDATE', () => {
    // Given
    const initState: any = {
      myKey: 'val',
      noKey: true
    }
    const action = { type: UPDATE_META, myKey: 'yo', secondKey: true }

    // When
    const nextState = reducer(initState, action)

    // Then
    expect(nextState).toMatchSnapshot()
  })
})
