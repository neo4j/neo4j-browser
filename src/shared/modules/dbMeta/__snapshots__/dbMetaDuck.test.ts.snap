// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`hydrating state can CLEAR to reset state 1`] = `
Object {
  "countAutomaticRefresh": Object {
    "enabled": true,
    "loading": false,
  },
  "databases": Array [],
  "functions": Array [],
  "labels": Array [],
  "nodes": 0,
  "procedures": Array [],
  "properties": Array [],
  "relationshipTypes": Array [],
  "relationships": 0,
  "role": null,
  "server": Object {
    "edition": null,
    "storeSize": null,
    "version": null,
  },
  "serverConfigDone": false,
  "settings": Object {
    "allowOutgoingConnections": false,
    "allowTelemetry": true,
    "authEnabled": true,
    "credentialTimeout": 0,
    "metricsNamespacesEnabled": false,
    "metricsPrefix": "neo4j",
    "postConnectCmd": "",
    "remoteContentHostnameAllowlist": "guides.neo4j.com, localhost",
    "retainConnectionCredentials": false,
    "retainEditorHistory": false,
  },
  "trialStatus": Object {
    "status": "unknown",
  },
}
`;

exports[`hydrating state can update meta values with UPDATE 1`] = `
Object {
  "myKey": "yo",
  "noKey": true,
  "secondKey": true,
}
`;

exports[`hydrating state can update server info 1`] = `
Object {
  "server": Object {
    "edition": "enterprise",
    "storeId": "xxxx",
    "version": "3.2.0-RC2",
  },
  "shouldKeep": true,
}
`;

exports[`hydrating state can update server settings 1`] = `
Object {
  "settings": Object {
    "browser.test": 1,
  },
  "shouldKeep": true,
}
`;

exports[`hydrating state should merge inital state and state on load 1`] = `
Object {
  "countAutomaticRefresh": Object {
    "enabled": true,
    "loading": false,
  },
  "databases": Array [],
  "foo": "bar",
  "functions": Array [],
  "labels": Array [],
  "nodes": 0,
  "procedures": Array [],
  "properties": Array [],
  "relationshipTypes": Array [],
  "relationships": 0,
  "role": null,
  "server": Object {
    "edition": null,
    "storeSize": null,
    "version": null,
  },
  "serverConfigDone": false,
  "settings": Object {
    "allowOutgoingConnections": false,
    "allowTelemetry": true,
    "authEnabled": true,
    "credentialTimeout": 0,
    "metricsNamespacesEnabled": false,
    "metricsPrefix": "neo4j",
    "postConnectCmd": "",
    "remoteContentHostnameAllowlist": "guides.neo4j.com, localhost",
    "retainConnectionCredentials": false,
    "retainEditorHistory": false,
  },
  "trialStatus": Object {
    "status": "unknown",
  },
}
`;
